rasterio/vfs.py

```diff
@@ -14,13 +16,17 @@ def parse_path(path, vfs=None):
     """Parse a file path or Apache VFS URL into its parts."""
     archive = scheme = None
     if vfs:
-        parts = vfs.split("://")
-        scheme = parts.pop(0) if parts else None
-        archive = parts.pop(0) if parts else None
+        parts = parse(vfs)
+        scheme = parts['scheme']
+        archive = parts['path']
+        if parts['authority'] and parts['authority'] != 'localhost':
+            archive = parts['authority'] + archive
     else:
-        parts = path.split("://")
-        path = parts.pop() if parts else None
-        scheme = parts.pop() if parts else None
+        parts = parse(path)
+        scheme = parts['scheme']
+        path = parts['path']
+        if parts['authority'] and parts['authority'] != 'localhost':
+            path = parts['authority'] + path
```

Comment:
@sgillies Any reason not to use Python's builtin `urllib.parse.urlparse()` for parsing?  It's `urlparse.urlparse()` in Python 2.

```python
import rfc3987
from urllib.parse import urlparse

url = 's3://landsat-pds/L8/139/045/LC81390452014295LGN00/LC81390452014295LGN00_B1.TIF'

u = urlparse(url)
r = rfc3987.parse(url)

assert u.scheme == r['scheme'])
assert u.netloc == r['authority'])
assert u.path == r['path'])
```

Here are the objects both return:

```python
ParseResult(scheme='s3', netloc='landsat-pds', path='/L8/139/045/LC81390452014295LGN00/LC81390452014295LGN00_B1.TIF', params='', query='', fragment='')
{'fragment': None, 'path': '/L8/139/045/LC81390452014295LGN00/LC81390452014295LGN00_B1.TIF', 'query': None, 'authority': 'landsat-pds', 'scheme': 's3'}
```